packages/enzyme/src/ShallowWrapper.js

```diff
@@ -268,65 +269,41 @@ class ShallowWrapper {
         const state = instance.state;
         const prevProps = instance.props || this[UNRENDERED].props;
         const prevContext = instance.context || this[OPTIONS].context;
-        const nextProps = { ...prevProps, ...props };
         const nextContext = context || prevContext;
         if (context) {
           this[OPTIONS] = { ...this[OPTIONS], context: nextContext };
         }
         this[RENDERER].batchedUpdates(() => {
+          // When shouldComponentUpdate returns false we shouldn't call componentDidUpdate.
+          // so we spy shouldComponentUpdate to know the result.
           let shouldRender = true;
-          // dirty hack:
-          // make sure that componentWillReceiveProps is called before shouldComponentUpdate
-          let originalComponentWillReceiveProps;
+          let spy;
           if (
             !this[OPTIONS].disableLifecycleMethods &&
             instance &&
-            typeof instance.componentWillReceiveProps === 'function'
+            typeof instance.shouldComponentUpdate === 'function'
           ) {
-            instance.componentWillReceiveProps(nextProps, nextContext);
-            originalComponentWillReceiveProps = instance.componentWillReceiveProps;
-            instance.componentWillReceiveProps = () => {};
+            spy = sinon.spy(instance, 'shouldComponentUpdate');
+          }
+          if (props) this[UNRENDERED] = cloneElement(adapter, this[UNRENDERED], props);
+          this[RENDERER].render(this[UNRENDERED], nextContext);
+          if (spy) {
+            shouldRender = spy.getCall(0).returnValue;
```

Comment:
Should we care about calling `shouldComponentUpdate` multiple times?

Reply:
Maybe it should get the last call, not the first?